Desktop: Focus editor after inserting datetime

CalebJohn opened the pull request:
fixes #4036 

Attach file and insert datetime were not returning focus to codemirror after being pressed.

laurent22 commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/CodeMirror.tsx:
```diff
@@ -182,13 +182,17 @@ function CodeMirror(props: NoteBodyEditorProps, ref: any) {
 							}
 						},
-						insertText: (value: any) => editorRef.current.insertAtCursor(value),
+						insertText: (value: any) => {
+							editorRef.current.insertAtCursor(value);
+							editorRef.current.focus();
+						},
 						attachFile: async () => {
 							const cursor = editorRef.current.getCursor();
 							const pos = cursorPositionToTextOffset(cursor, props.content);
 
 							const newBody = await commandAttachFileToBody(props.content, null, { position: pos });
 							if (newBody) editorRef.current.updateBody(newBody);
+							editorRef.current.focus();
```
I would prefer commands not have side-effects, especially since they are used from plugins (I can imagine some cases where a plugin might change text but without wanting to focus the editor). Instead is it not possible to add the behaviour to the toolbar button click event? i.e. every time any button is clicked, we focus the editor, although I guess there should be an exception for some buttons, like Back/Forward and External Edit. What do you think?

CalebJohn replied:
That makes sense to me, I was viewing these commands as the toolbar button presses but now that we have a plugin system that is an incorrect assumption. The `wrapSelectionWithStrings` function also focuses the editor (it's used by a few of the commands), so I'll remove that as well to enforce consistency.

CalebJohn commented:
@laurent22 I've added the behaviour you described, and I've added it as optional so the default isn't to focus the editor. It might be better as an interface to have that inverted however, i.e. should toolbar buttons automatically pass focus to the editor unless explicitly disabled?

laurent22 commented:
Hmm, thanks that would work but based on your code all the toolbar button commands have the `focusAfterPress` set to `true` and all those that are not have it set to `false`. So how about removing the property, and in the button onClick handler, always call `this.service.execute('focus')`?

CalebJohn commented:
> So how about removing the property, and in the button onClick handler, always call this.service.execute('focus')?

The note back/forward buttons don't focus after press, and the external editor doesn't focus after press. In general I think it's sane that everything on the editor toolbar should focus the editor. But I think we'll find that plugin creators will use the toolbar for creative interactions that don't require focusing the editor. So for that reason I think we need to have the option to focus the editor.

I'm happy to switch the default to focusing the editor if that's what you'd prefer. Just let me know.

laurent22 commented:
Sorry you're right, I must have been tired yesterday since we've obviously already discussed this.

I'm a bit annoyed with seeing text editor related properties directly in the command definition. If we really have to add a new property, and it seems we do, I wonder if we can express that in a better way. What distinguishes the back/forward and external edit commands from all the other commands basically? Maybe it's just that they don't change text hence they shouldn't focus anything. Or they are simply not really text editor commands, but we've put them here for convenience. Not sure, but if you have any suggestion please let me know.

Also all commands are global that it should be "editor.focus" (I've started adding a namespace to commands)

CalebJohn commented:
> I'm a bit annoyed with seeing text editor related properties directly in the command definition.

I know what you mean, it's a slippery slope of adding unrelated flags and features. My personal leaning is to have all the commands that interact with the editor focus it (as was originally done for this PR). The reasoning being that it makes the most intuitive sense, e.g. whenever you edit text you are required to be focused on the editor, so why should it be different when it is done via command? 

I'm not married to that, and I could see arguments for other behaviours, but that is my preference.
In terms of ergonomics for the optional focus, It probably does make sense to have focus the default and allow toolbar buttons to disable it. We could also call the property something like `textButton` or `editorButtton` and only those would focus the editor.

laurent22 commented:
Hmm, yes perhaps we indeed only need to had the focus call in the command as you had at the beginning.

The reason I preferred to keep the focus separate is if a plugin somehow wants to change the note body dynamically from a webview. Let's say someone create a quick note plugin that parses the user's input as they type, and add info to the note. For example they type this:

`Meeting with @alice and @bob tomorrow`

The plugin automatically create a new alarm for tomorrow, and then fetches the contact info from somewhere and adds this to the body as the user types:

```markdown
## Participants

- Alice +012345678 alice@corp.com
- Bob +012345678 bob@corp.com
```

If the focus jumps to the body, that kind of plugin would be hard to make because the user would have to refocus the text input every time the body is updated.

Then again, no-one is trying to make such plugin, so it's possible it's all YAGNI, and since we don't have any non-complicated solution, the simple fix you had first might be the way to go for now.

CalebJohn commented:
You bring up a good point, and I think in this specific case of trying to support a plugin ecosystem it makes sense to not shoehorn use cases. I think your example would make a really useful plugin and we should make sure it can be created. I can even imagine that people might eventually make a plugin that replaces the editor with Ace, or Monaco or whatever they want, so it makes sense to allow some independence.

I'll think about this a bit more, but right now I'm leaning towards categorizing buttons as editor buttons (maybe insert buttons) and independent buttons. Editor buttons would focus the editor and others would not.

laurent22 commented:
> I'll think about this a bit more, but right now I'm leaning towards categorizing buttons as editor buttons (maybe insert buttons) and independent buttons. Editor buttons would focus the editor and others would not.

Indeed maybe something like this, and in fact it's related to the "editor" namespace I've started adding to the editor commands. If we can identify that a command is for the editor, by checking that it starts with "editor.", then we can easily implement the focus logic. And in that case, we wouldn't need adding a new property to the command declaration.

laurent22 commented:
I thought about this a bit more and I think that's what we should do:

- New editor commands will be prefixed with "editor." so we'll know they are editor commands
- Old editor commands are not like this, and I found that changing their names would be quite complex, especially since users might have associated shortcuts with these old commands. So we'll keep the old commands as they are.
- I'd prefer not to add a new property to the command declaration just to support these old commands, so instead I'd suggest we simply create a new function `isEditorCommand`, which will just check if the command is one of "textBold", "textItalic", etc. or starts with "editor." and returns `true` in that case. That way we handle both new and old commands without for now changing the command system. Later if we find it becomes necessary, we can always add that property.

Can you think of any potential issue with this approach?

CalebJohn commented:
That sounds good to me. My only concern is supporting the above workflow. In this system would it be possible to set the note text without triggering focus on the editor? Or would we want to create an `editor.setText` and a `setText` command to get the difference in focus behaviour?

laurent22 commented:
>  In this system would it be possible to set the note text without triggering focus on the editor?

Yes, in the toolbar button onClick handler, you'd have what you have now, but instead of checking for `command.declaration.focusAfterPress`, you'd call `isEditorCommand(commandName)`. I think that would work?

CalebJohn commented on packages/lib/services/commands/ToolbarButtonUtils.ts:
```diff
@@ -52,8 +70,11 @@ export default class ToolbarButtonUtils {
 			iconName: command.declaration.iconName,
 			enabled: newEnabled,
-			onClick: () => {
+			onClick: async () => {
```
This async was added by the the pre-commit hook

laurent22 replied:
I'm not sure it's actually the pre-commit hook, maybe it's when you rebased on dev? In any case this is fine. I've added a TypeScript so that all promises must either be awaited or explicitly marked as "void" (as the call to service.execute below).

CalebJohn commented:
@laurent22 Thanks for that, took me awhile to catch up :sweat_smile:.

I've now updated this pull to do exactly what you suggested. The only command this doesn't work for is `textLink` because it needs the focus to go to the dialogue, and then return to the editor after. So for that one I put the focus code in `CodeMirror.tsx`, I think it's a fair compromise.

laurent22 commented on packages/lib/services/commands/ToolbarButtonUtils.ts:
```diff
@@ -33,6 +33,24 @@ export default class ToolbarButtonUtils {
+		return (commandName.indexOf('editor.') === 0 ||
+				// These commands are grandfathered in, but in the future
+				// all editor commands should start with "editor."
+				commandName === 'textBold' ||
+				commandName === 'textItalic' ||
+				commandName === 'textCode' ||
+				commandName === 'attachFile' ||
+				commandName === 'textNumberedList' ||
+				commandName === 'textBulletedList' ||
+				commandName === 'textCheckbox' ||
+				commandName === 'textHeading' ||
+				commandName === 'textHorizontalRule' ||
+				commandName === 'insertDateTime'
```
It looks like a few of them are missing, like textLink, textCopy, etc. shouldn't they be added too? For info, the full list is in editorCommandDeclaration.ts

CalebJohn replied:
Thanks, that was an oversight. I actually copied these commands out of editorCommandDeclaration.ts but missed the top few by mistake.

laurent22 commented on packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/CodeMirror.tsx:
```diff
@@ -170,6 +168,7 @@ function CodeMirror(props: NoteBodyEditorProps, ref: any) {
 						textItalic: () => wrapSelectionWithStrings('*', '*', _('emphasised text')),
 						textLink: async () => {
 							const url = await dialogs.prompt(_('Insert Hyperlink'));
+							editorRef.current.focus();
 							if (url) wrapSelectionWithStrings('[', `](${url})`);
```
Any reason why focus is hard-coded here?

CalebJohn replied:
This is the definition of the `textLink` command. If this command was included as an editor command then the popup box wouldn't be focused and the editor would get the focus instead. I thought it was okay for this command (and any other like it) to be a special case of focus, because they are more complex in regards to handling focus i.e. they need to move focus around, not just on the editor.

laurent22 commented on packages/lib/services/commands/ToolbarButtonUtils.ts:
```diff
@@ -33,6 +33,30 @@ export default class ToolbarButtonUtils {
+		return (commandName.indexOf('editor.') === 0 ||
+				// These commands are grandfathered in, but in the future
+				// all editor commands should start with "editor."
+				// WARNING: Some commands such as textLink are not defined here
+				// because they are more complex and handle focus manually
```
Thanks for adding this comment, it's always good to know that when looking at the code a few months later.

CalebJohn replied:
And thanks for all your help on this!

laurent22 reviewed (approved):
Looks good now, thanks @CalebJohn!
